pulsar-broker/src/test/java/org/apache/pulsar/client/api/BatchedProducerConsumerTest.java

```diff
@@ -0,0 +1,138 @@
+        super.internalCleanup();
+    }
+
+    @Override
+    protected void doInitConf() {
+        conf.setAcknowledgmentAtBatchIndexLevelEnabled(true);
+    }
+
+    @Test
+    public void testNegativeAckAndLongAckDelayWillNotLeadRepeatConsume() throws Exception {
+        final String topicName = BrokerTestUtil.newUniqueName("persistent://public/default/tp_");
+        final String subscriptionName = "s1";
+        final int redeliveryDelaySeconds = 2;
+
+        // Create producer and consumer.
+        Producer<String> producer = pulsarClient.newProducer(Schema.STRING)
+                .topic(topicName)
+                .enableBatching(true)
+                .batchingMaxMessages(1000)
+                .batchingMaxPublishDelay(1, TimeUnit.HOURS)
+                .create();
+        ConsumerImpl<String> consumer = (ConsumerImpl<String>) pulsarClient.newConsumer(Schema.STRING)
+                .topic(topicName)
+                .subscriptionName(subscriptionName)
+                .subscriptionType(SubscriptionType.Shared)
+                .negativeAckRedeliveryDelay(redeliveryDelaySeconds, TimeUnit.SECONDS)
+                .enableBatchIndexAcknowledgment(true)
+                .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+                .acknowledgmentGroupTime(1, TimeUnit.HOURS)
+                .subscribe();
+
+        // Send 10 messages in batch.
+        ArrayList<String> messagesSent = new ArrayList<>();
+        List<CompletableFuture<MessageId>> sendTasks = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            String msg = Integer.valueOf(i).toString();
+            sendTasks.add(producer.sendAsync(Integer.valueOf(i).toString()));
+            messagesSent.add(msg);
+        }
+        producer.flush();
+        FutureUtil.waitForAll(sendTasks);
+
+        // Receive messages.
+        ArrayList<String> messagesReceived = new ArrayList<>();
+        // NegativeAck "batchMessageIdIndex1" once.
+        boolean index1HasBeenNegativeAcked = false;
+        while (true) {
+            Message<String> message = consumer.receive(2, TimeUnit.SECONDS);
+            if (message == null) {
+                break;
+            }
+            if (index1HasBeenNegativeAcked) {
+                messagesReceived.add(message.getValue());
+                consumer.acknowledge(message);
+                continue;
+            }
+            if (message.getMessageId() instanceof BatchMessageIdImpl batchMessageId) {
+                if (batchMessageId.getBatchIndex() == 1) {
+                    consumer.negativeAcknowledge(message);
+                    index1HasBeenNegativeAcked = true;
+                    continue;
+                }
+            }
+            messagesReceived.add(message.getValue());
+            consumer.acknowledge(message);
+        }
+
+        // Wait the message negative acknowledgment finished.
+        Thread.sleep(redeliveryDelaySeconds * 3);
```

Comment:
`Thread.sleep` accepts milliseconds, not seconds.

BTW, is there a better way to avoid sleeping? e.g. wait for some condition?

Reply:
fixed